x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/topmetrics/InternalTopMetrics.java

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.analytics.topmetrics;
+
+import org.elasticsearch.common.io.stream.NamedWriteable;
+import org.elasticsearch.common.io.stream.NamedWriteableRegistry;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class InternalTopMetrics extends InternalAggregation {
+    public static List<NamedWriteableRegistry.Entry> writeables() {
+        return Arrays.asList(
+                new NamedWriteableRegistry.Entry(SortValue.class, DoubleSortValue.NAME, DoubleSortValue::new),
+                new NamedWriteableRegistry.Entry(SortValue.class, LongSortValue.NAME, LongSortValue::new));
+    }
+
+    private final DocValueFormat sortFormat;
+    private final SortOrder sortOrder;
+    private final SortValue sortValue;
+    private final String metricName;
+    private final double metricValue;
+
+    private InternalTopMetrics(String name, DocValueFormat sortFormat, SortOrder sortOrder, SortValue sortValue, String metricName,
+            double metricValue, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        this.sortFormat = sortFormat;
+        this.sortOrder = sortOrder;
+        this.sortValue = sortValue;
+        this.metricName = metricName;
+        this.metricValue = metricValue;
+    }
+
+    InternalTopMetrics(String name, DocValueFormat sortFormat, SortOrder sortOrder, Object sortValue, String metricName,
```

Comment:
Oh I see, we don't actually know the SortValue.  BucketedSort only gives us back an object, because we don't know which "ForFoo" was used for the sort value (`ForLong`, etc).

Hmm.  I wonder if we should do this conversion on the other end?  Something like `toSortValue()` on `BucketedSort`, which each implementation knows how to turn into an implementation of SortValue and then serialize that?

It feels a little cleaner since we don't need instance-of checks to resolve the SortValue, and would protect against situations where two types serialize the same data type  but should be interpreted as something different on the receiving end.  I think the various Bytes implementations might run into this (string bytes != range bytes for example).  Dunno if we'll ever implement situations where we'd run into it though.

OTOH, it adds extra serialization code to all the SortValues.

Reply:
I think we could do the conversion on the other end if we want `SortValue` to become a "core" thing. I think it'd be cleaner, even. Do you think I should move `SortValue` over?

Reply:
Ahhh right, forgot about the split here. Hmm. I'm leaning towards "yes", feels generally like a "core" thing.  WDYT?

Reply:
Sure! I'll move it over.